src/main/kotlin/org/rust/ide/typing/paste/JsonStructParser.kt

```diff
@@ -136,14 +139,59 @@ private class StructParser {
             foundDataTypes.add(dataType)
         }
 
-        val containsNull = foundDataTypes.any { it is DataType.Nullable }
+        return DataType.Array(generateContainedType(foundDataTypes))
+    }
+
+    private fun extractStructType(type: DataType): DataType.StructRef? {
+        return when {
+            type is DataType.StructRef -> type
+            type is DataType.Nullable && type.type is DataType.StructRef -> type.type
+            else -> null
+        }
+    }
+
+    private fun generateContainedType(types: Set<DataType>): DataType {
+        val containsNull = types.any { it is DataType.Nullable }
+        val structTypes = types.mapNotNull { extractStructType(it) }
         val innerType = when {
-            foundDataTypes.size == 1 -> foundDataTypes.first()
-            foundDataTypes.size == 2 && containsNull -> DataType.Nullable(foundDataTypes.first())
+            types.size == 1 -> types.first()
+            types.size == 2 && containsNull -> DataType.Nullable(types.first { it !is DataType.Nullable })
+            structTypes.size == types.size -> {
+                val type = unifyStructs(structTypes)
+                if (containsNull) {
+                    DataType.Nullable(type)
+                } else {
+                    type
+                }
+            }
+            containsNull -> DataType.Nullable(DataType.Unknown)
             else -> DataType.Unknown
         }
+        return innerType
+    }
+
+    /**
+     * Look through a list of structs and try to unify them to a single struct type with optional fields.
+     */
+    private fun unifyStructs(structTypes: List<DataType.StructRef>): DataType {
+        // Gather all field keys
+        val foundFields = structTypes
+            .flatMap { it.struct.fields.entries }
+            .groupByTo(mutableMapOf(), { it.key }, { it.value })
```

Comment:
```suggestion
            .groupBy({ it.key }, { it.value })
```

Reply:
I need the return type to be `MutableMap<Key, MutableList<...>>`, otherwise it won't be able to add the nullable type on line 185.